File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/152_Maximum_Product_Subarray.swift
import Foundation

class Solution {
    func maxProduct(_ nums: [Int]) -> Int {
        var maxProd = 1
        var minProd = 1
        var best = nums[0]
        for i in 0 ..< nums.count {
            var currentNum = nums[i]
            if currentNum < 0 {
                swap(&maxProd, &minProd)
            }
            maxProd = max(maxProd * currentNum, currentNum)
            minProd = min(minProd * currentNum, currentNum)
            best = max(best, maxProd)
        }
        return best
    }
}

File: leetcode.com/python/152_Maximum_Product_Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        minProd, maxProd, bestSoFar = 1, 1, nums[0]
        for i in range(len(nums)):
            currentNum = nums[i]
            if currentNum < 0:
                minProd, maxProd = maxProd, minProd
            maxProd = max(maxProd*currentNum, currentNum)
            minProd = min(minProd*currentNum, currentNum)
            bestSoFar = max(bestSoFar, maxProd)
        return bestSoFar
